### PR TITLE
Custom confirmation logic support

### PR DESCRIPTION
**Pull request by clementetb**
Add support for calling the custom confirmation function from the SDKs

### CODE REVIEW

File: src/sync/app.cpp [hunk 1]
```diff
@@ -228,6 +228,29 @@ void App::UsernamePasswordProviderClient::resend_confirmation_email(const std::s
     }, handler);
 }
 
+void App::UsernamePasswordProviderClient::retry_custom_confirmation(const std::string& email,
```
Comment (jsflax) [inline src/sync/app.cpp:231]: Is this the correct function name? Maybe I'm misunderstanding, but it seems like `call_custom_confirmation_function` might be a bit clearer.

Comment (cmelchior) [reply src/sync/app.cpp]: Sounds like we need to agree on how to expose this to users 😅 https://github.com/realm/realm-java/pull/7079#discussion_r484393051

File: src/sync/app.hpp [hunk 2]
```diff
@@ -188,6 +188,12 @@ class App : public std::enable_shared_from_this<App>, public AuthRequestClient,
         void send_reset_password_email(const std::string& email,
                                        std::function<void(util::Optional<AppError>)> completion_block);
 
+        /// Calls the custom confirmation function on a user for a given email.
+        /// @param email The email address of the user to call the custom confirmation for.
+        /// @param completion_block A callback to be invoked once the call is complete.
+        void call_custom_confirmation_function(const std::string& email,
```
[the remark this answers is not in the document]
Comment (ironage) [reply src/sync/app.hpp:194]: You can add a custom confirmation function to the config of the docker test server in the same way that we test email registration by only allowing emails that contain "realm_tests_do_autoverify" (see [here](https://github.com/realm/realm-object-store/blob/v10/tests/mongodb/functions/confirmFunc/source.js#L44) and [here](https://github.com/realm/realm-object-store/blob/v10/tests/sync/app.cpp#L316))

[the remark this answers is not in the document]
Comment (clementetb) [reply src/sync/app.hpp]: Have added test cases for the custom confirmation function, following the same pattern as `call_reset_password_function`.

**Review by ironage — APPROVED** [review]

File: src/sync/app.cpp (hunk 1 above)
Comment (cmelchior) [reply src/sync/app.cpp:231]: By popular vote, we agreed on `retry_custom_confirmation`

**Review by cmelchior — APPROVED** [review]

**Pull request merged**